src/Carousel.js

```diff
@@ -100,441 +124,430 @@ const propTypes = {
     }
+  });
 
-    this.select(index, e, 'next');
-  };
+  const slideDirection = direction === 'next' ? 'left' : 'right';
 
-  handlePrev = e => {
-    if (this._isSliding) return;
+  useUpdateEffect(() => {
+    if (slide) {
+      // These callbacks will be handled by the <Transition> callbacks.
+      return;
+    }
 
-    const { wrap, activeIndex } = this.props;
+    if (onSlide) {
+      onSlide(renderedActiveIndex, slideDirection);
+    }
+    if (onSlid) {
+      onSlid(renderedActiveIndex, slideDirection);
+    }
+  }, [renderedActiveIndex]);
 
-    let index = activeIndex - 1;
+  const orderClassName = `${prefix}-item-${direction}`;
+  const directionalClassName = `${prefix}-item-${slideDirection}`;
 
-    if (index < 0) {
-      if (!wrap) return;
-      index = countChildren(this.props.children) - 1;
-    }
+  const handleEnter = useCallback(
+    node => {
+      triggerBrowserReflow(node);
 
-    this.select(index, e, 'prev');
-  };
+      if (onSlide) {
+        onSlide(renderedActiveIndex, slideDirection);
+      }
+    },
+    [onSlide, renderedActiveIndex, slideDirection],
+  );
 
-  safeSetState(state, cb) {
-    if (this.isUnmounted) return;
-    this.setState(state, () => !this.isUnmounted && cb());
-  }
+  const handleEntered = useCallback(() => {
+    setIsSliding(false);
 
-  // This might be a public API.
-  pause() {
-    this._isPaused = true;
-    clearInterval(this._interval);
-    this._interval = null;
-  }
+    if (onSlid) {
+      onSlid(renderedActiveIndex, slideDirection);
+    }
+  }, [onSlid, renderedActiveIndex, slideDirection]);
+
+  const handleKeyDown = useCallback(
+    event => {
+      if (keyboard && !/input|textarea/i.test(event.target.tagName)) {
+        switch (event.key) {
+          case 'ArrowLeft':
+            event.preventDefault();
+            prev(event);
+            return;
+          case 'ArrowRight':
+            event.preventDefault();
+            next(event);
+            return;
+          default:
+        }
+      }
+
+      if (onKeyDown) {
+        onKeyDown(event);
+      }
+    },
+    [keyboard, onKeyDown, prev, next],
+  );
 
-  cycle() {
-    this._isPaused = false;
+  const [pausedOnHover, setPausedOnHover] = useState(false);
 
-    clearInterval(this._interval);
-    this._interval = null;
+  const handleMouseOver = useCallback(
+    event => {
+      if (pause === 'hover') {
+        setPausedOnHover(true);
+      }
 
-    if (this.props.interval && !this._isPaused) {
-      this._interval = setInterval(
-        document.visibilityState ? this.handleNextWhenVisible : this.handleNext,
-        this.props.interval,
-      );
-    }
-  }
+      if (onMouseOver) {
+        onMouseOver(event);
+      }
+    },
+    [pause, onMouseOver],
+  );
 
-  to(index, event) {
-    const { children } = this.props;
-    if (index < 0 || index > countChildren(children) - 1) {
-      return;
-    }
+  const handleMouseOut = useCallback(
+    event => {
+      setPausedOnHover(false);
```

Comment:
similar to `handleMouseOver` in case `pause` is `false`, this will avoid extra rerender
```
if (pause === 'hover') {
  setPausedOnHover(false);
}
```
but I guess you did it for a reason and there might be an edge case that I am not seeing :)

Reply:
`setState` for a value that's the same value (`===`) won't cause another render when it's done in a callback, I think.

It _will_ cause an active render function to bail out, but that's not at issue with callbacks.